src/dmd/cppmangle.d

```diff
@@ -1373,4 +1453,400 @@ extern(C++):
+            {
+                this.context.res = (*analyzed_ti.tiargs)[idx + t.tiargs.dim];
+
+                if (auto ttp = arg.isTemplateTypeParameter())
+                    ttp.defaultType.accept(this);
+                else if (auto tvp = arg.isTemplateValueParameter())
+                    tvp.defaultValue.accept(this);
+                else if (auto tvp = arg.isTemplateThisParameter())
+                    tvp.defaultType.accept(this);
+                else if (auto tvp = arg.isTemplateAliasParameter())
+                    tvp.defaultAlias.visitObject(this);
+                else
+                    assert(0, arg.toString());
+            }
+        }
+        buf.writeByte('E');
+        if (needclosing)
+            buf.writeByte('E');
+    }
+
+    /// Ditto
+    override void visit(IntegerExp t)
+    {
+        this.buf.writeByte('L');
+        t.type.accept(this);
+        this.buf.print(t.getInteger());
+        this.buf.writeByte('E');
+    }
+
+    override void visit(Nspace t)
+    {
+        if (auto p = getQualifier(t))
+            p.accept(this);
+
+        if (isStd(t))
+            buf.writestring("St");
+        else
+        {
+            this.writeIdentifier(t.ident);
+            this.append(t);
+        }
+    }
+
+    override void visit(Type t)
+    {
+        error(t);
+    }
+
+    void visit(Tuple t)
+    {
+        assert(0);
+    }
+
+    /**
+     * Helper function to go through the `TemplateParameter`s
+     * and perform a substitution, if possible.
+     *
+     * Returns:
+     *   The `TemplateParameter` that was used, or `null`.
+     */
+    private bool writeTemplateSubstitution (
+        const ref Identifier ident, TemplateParameters* params, Type type)
+    {
+        foreach (idx, param; *params)
+        {
+            if (param.ident == ident)
+            {
+                if (type)
+                    CV_qualifiers(type);
+                if (this.substitute(param))
+                    return true;
+                this.append(param);
+
+                // Since we didn't run semantic yet, all we get are
+                // `TypeIdentifier`, but we need to check if it's an expression
+                // or whatnot
+                if (param.isTemplateValueParameter())
+                    buf.writeByte('X');
+                buf.writeByte('T');
+                writeSequenceFromIndex(idx);
+                buf.writeByte('_');
+                if (param.isTemplateValueParameter())
+                    buf.writeByte('E');
+                return true;
+            }
+        }
+        return false;
+    }
+
+    /**
+     * Given a template instance `t`, write it's qualified name
+     * without the template parameter list
+     *
+     * Params:
+     *   t = Post-parsing `TemplateInstance` pointing to the symbol
+     *       to mangle (one level deep)
+     *
+     * Returns:
+     *   `true` if the name was qualified and need an ending `E`
```

Comment:
and needs an ending `E`.

You could also say requires here, need sounds urgent. :-)